feat(firebase-admin): upgrade to v11

rapzo opened the pull request:
Started in #8144

jtoar commented:
Thanks @rapzo! Gotta love the one line fix. To add context, without this line, I get this error in the terminal:

```
api | 20:46:01 🚨 Exception in getAuthenticationContext: The default Firebase app does not exist. Make sure you call initializeApp() before using any of the Firebase services. 
api | 
api | 🚨 Error Info
api | 
api | {}
api |  
api | 🥞 Error Stack
api | 
api | Error: Exception in getAuthenticationContext: The default Firebase app does not exist. Make sure you call initializeApp() before using any of the Firebase services.
api |     at onContextBuilding (/Users/dom/prjcts/redwood/redwood-app/node_modules/@redwoodjs/graphql-server/dist/plugins/useRedwoodAuthContext.js:32:15)
api |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
api |     at async Object.contextFactory (/Users/dom/prjcts/redwood/redwood-app/node_modules/@envelop/core/cjs/orchestrator.js:198:45)
api |     at async processRequest (/Users/dom/prjcts/redwood/redwood-app/node_modules/graphql-yoga/cjs/process-request.js:46:26)
api |     at async YogaServer.getResultForParams (/Users/dom/prjcts/redwood/redwood-app/node_modules/graphql-yoga/cjs/server.js:331:26)
api |     at async YogaServer.handle (/Users/dom/prjcts/redwood/redwood-app/node_modules/graphql-yoga/cjs/server.js:78:29)
api |     at async handlerWithErrorHandling (/Users/dom/prjcts/redwood/redwood-app/node_modules/@whatwg-node/server/cjs/plugins/useErrorHandling.js:33:38)
api |     at async handleRequest (/Users/dom/prjcts/redwood/redwood-app/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:73:24)
api |     at async handlerFn (/Users/dom/prjcts/redwood/redwood-app/node_modules/@redwoodjs/graphql-server/dist/functions/graphql.js:182:24)
api |     at async execFn (/Users/dom/prjcts/redwood/redwood-app/node_modules/@redwoodjs/graphql-server/dist/functions/graphql.js:237:16)
```

It looks like we need to make one more change to the api-side template. (And write a codemod to go with it, but I can handle that part.) With this PR, I get this warning (I think it's a warning at least, but it looks like a stack trace which I normally associate with errors):

```
api | /Users/dom/prjcts/redwood/redwood-app/node_modules/firebase-admin/lib/app/lifecycle.js:41
api |                 throw new error_1.FirebaseAppError(error_1.AppErrorCodes.DUPLICATE_APP, 'The default Firebase app already exists. This means you called initializeApp() ' +
api |                       ^
api | 
api | FirebaseAppError: The default Firebase app already exists. This means you called initializeApp() more than once without providing an app name as the second argument. In most cases you only need to call initializeApp() once. But if you do want to initialize multiple apps, pass a second argument to initializeApp() to give each app a unique name.
api |     at AppStore.initializeApp (/Users/dom/prjcts/redwood/redwood-app/node_modules/firebase-admin/lib/app/lifecycle.js:41:23)
api |     at FirebaseNamespaceInternals.initializeApp (/Users/dom/prjcts/redwood/redwood-app/node_modules/firebase-admin/lib/app/firebase-namespace.js:43:35)
api |     at FirebaseNamespace.initializeApp (/Users/dom/prjcts/redwood/redwood-app/node_modules/firebase-admin/lib/app/firebase-namespace.js:246:30)
api |     at Object.<anonymous> (/Users/dom/prjcts/redwood/redwood-app/api/src/lib/auth.ts:6:24)
api |     at Module._compile (node:internal/modules/cjs/loader:1254:14)
api |     at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
api |     at Module.load (node:internal/modules/cjs/loader:1117:32)
api |     at Module._load (node:internal/modules/cjs/loader:958:12)
api |     at Module.require (node:internal/modules/cjs/loader:1141:19)
api |     at require (node:internal/modules/cjs/helpers:110:18) {
api |   errorInfo: {
api |     code: 'app/duplicate-app',
api |     message: 'The default Firebase app already exists. This means you called initializeApp() more than once without providing an app name as the second argument. In most cases you only need to call initializeApp() once. But if you do want to initialize multiple apps, pass a second argument to initializeApp() to give each app a unique name.'
api |   },
api |   codePrefix: 'app'
api | }
```

That's because the template also initializes an app:

https://github.com/redwoodjs/redwood/blob/c7fe0a620e71071ba7fbaf8c62699d90e8c5cb7a/packages/auth-providers/firebase/setup/src/templates/api/lib/auth.ts.template#L5-L8

Let me check in with @Tobbe who did all the auth work next week.

Tobbe commented:
I just tried this and got the same issue @jtoar reported when I tried with the changes introduced in this PR.
But if I just undo these changes to `decoder.ts` everything start working for me

![image](https://github.com/redwoodjs/redwood/assets/30793/202a740c-afc8-4291-b195-aeff4c78c83c)

It's worth noting that testing changes to the auth packages is difficult. You need to get them into your test project, and make sure they're not overwritten/reverted by a `yarn install`... It's super easy to end up with mismatched versions of packages. I had to do some coaxing to get `firebase-admin` to be v11.8.0 everywhere. And when I finally fixed that one webpack for some reason decided to get out of sync and threw weird errors until I fixed that one too... `yarn why` is your friend here.

![image](https://github.com/redwoodjs/redwood/assets/30793/55c34576-5001-466e-bd7b-cdf3995c1ab2)

But @jtoar knows all of this, and is better with these kinds of issues than I am, so I'm still worried I've messed something up in my own testing.

This is the code I verified with
```tsx
import { useAuth } from 'src/auth'

const HomePage = () => {
  const { isAuthenticated, signUp } = useAuth()

  return (
    <>
      <h1>HomePage</h1>

      <p>{JSON.stringify({ isAuthenticated })}</p>
      <button
        onClick={() =>
          signUp({
            email: 'my-email@domain.com',
            password: 'super secret password',
          })
        }
      >
        sign up
      </button>
    </>
  )
}

export default HomePage
```
![image](https://github.com/redwoodjs/redwood/assets/30793/2221507b-44b6-4308-abc2-a2652185860b)

rapzo commented:
Interesting. So, the update didn't work because conflicting
`firebase-admin` versions?
Gonna do some more thorough tests then.
Thanks for the feedback!

On Sat, May 13, 2023, 12:14 Tobbe Lundberg ***@***.***> wrote:

> I just tried this and got the same issue @jtoar <https://github.com/jtoar>
> reported when I tried with the changes introduced in this PR.
> But if I just undo these changes to decoder.ts everything start working
> for me
>
> [image: image]
> <https://user-images.githubusercontent.com/30793/238130306-202a740c-afc8-4291-b195-aeff4c78c83c.png>
>
> It's worth noting that testing changes to the auth packages is difficult.
> You need to get them into your test project, and make sure they're not
> overwritten/reverted by a yarn install... It's super easy to end up with
> mismatched versions of packages. I had to do some coaxing to get
> firebase-admin to be v11.8.0 everywhere. And when I finally fixed that
> one webpack for some reason decided to get out of sync and threw weird
> errors until I fixed that one too... yarn why is your friend here.
>
> [image: image]
> <https://user-images.githubusercontent.com/30793/238130486-55c34576-5001-466e-bd7b-cdf3995c1ab2.png>
>
> But @jtoar <https://github.com/jtoar> knows all of this, and is better
> with these kinds of issues than I am, so I'm still worried I've messed
> something up in my own testing.
>
> This is the code I verified with
>
> import { useAuth } from 'src/auth'
> const HomePage = () => {
>   const { isAuthenticated, signUp } = useAuth()
>
>   return (
>     <>
>       <h1>HomePage</h1>
>
>       <p>{JSON.stringify({ isAuthenticated })}</p>
>       <button
>         onClick={() =>
>           signUp({
>             email: ***@***.***',
>             password: 'super secret password',
>           })
>         }
>       >
>         sign up
>       </button>
>     </>
>   )}
> export default HomePage
>
> [image: image]
> <https://user-images.githubusercontent.com/30793/238130632-2221507b-44b6-4308-abc2-a2652185860b.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redwoodjs/redwood/pull/8305#issuecomment-1546625028>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABECTGUFTLWTP3NDSWGW2LXF5UKFANCNFSM6AAAAAAX73D6NA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rapzo commented:
From what I understood from my tests is that the current `initializeApp`
call at the `app/api/src/lib/auth.ts` is too late a call in terms of
firebase initialization.
So, removing the `initializeApp` from the generated `auth.ts` file, but
keeping the line I added, makes it work without warnings on my side.
Updated the PR with that snippet removed.
Let me know if that makes sense or if I should dig for a solution somewhere
else in the project.

On Sat, 13 May 2023 at 12:22, Rui Pedro Lima ***@***.***>
wrote:

> Interesting. So, the update didn't work because conflicting
> `firebase-admin` versions?
> Gonna do some more thorough tests then.
> Thanks for the feedback!
>
> On Sat, May 13, 2023, 12:14 Tobbe Lundberg ***@***.***>
> wrote:
>
>> I just tried this and got the same issue @jtoar
>> <https://github.com/jtoar> reported when I tried with the changes
>> introduced in this PR.
>> But if I just undo these changes to decoder.ts everything start working
>> for me
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/30793/238130306-202a740c-afc8-4291-b195-aeff4c78c83c.png>
>>
>> It's worth noting that testing changes to the auth packages is difficult.
>> You need to get them into your test project, and make sure they're not
>> overwritten/reverted by a yarn install... It's super easy to end up with
>> mismatched versions of packages. I had to do some coaxing to get
>> firebase-admin to be v11.8.0 everywhere. And when I finally fixed that
>> one webpack for some reason decided to get out of sync and threw weird
>> errors until I fixed that one too... yarn why is your friend here.
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/30793/238130486-55c34576-5001-466e-bd7b-cdf3995c1ab2.png>
>>
>> But @jtoar <https://github.com/jtoar> knows all of this, and is better
>> with these kinds of issues than I am, so I'm still worried I've messed
>> something up in my own testing.
>>
>> This is the code I verified with
>>
>> import { useAuth } from 'src/auth'
>> const HomePage = () => {
>>   const { isAuthenticated, signUp } = useAuth()
>>
>>   return (
>>     <>
>>       <h1>HomePage</h1>
>>
>>       <p>{JSON.stringify({ isAuthenticated })}</p>
>>       <button
>>         onClick={() =>
>>           signUp({
>>             email: ***@***.***',
>>             password: 'super secret password',
>>           })
>>         }
>>       >
>>         sign up
>>       </button>
>>     </>
>>   )}
>> export default HomePage
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/30793/238130632-2221507b-44b6-4308-abc2-a2652185860b.png>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/redwoodjs/redwood/pull/8305#issuecomment-1546625028>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AABECTGUFTLWTP3NDSWGW2LXF5UKFANCNFSM6AAAAAAX73D6NA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

-- 
Rui Pedro Lima

Tobbe commented:
I found some of my old notes from when I last touched this code

---

> I get an error in my terminal
![image](https://github.com/redwoodjs/redwood/assets/30793/0fa711c7-c5a6-421b-80f5-c2eb54500868)
> I get it when returning back to the app after logging in in the google oauth popup
> If I call initializeApp an extra time I get
![image](https://github.com/redwoodjs/redwood/assets/30793/a9197a8d-e4aa-45f3-8215-46b0bb6f6c4c)
> If I remove that extra call to initializeApp I get
![image](https://github.com/redwoodjs/redwood/assets/30793/60c824fb-f92a-4ac3-bc01-e118237f4a42)
> I just can't win 😅 

> Got it working by putting all the `authDecoder` code inside the project instead of using the implementation that's exported by `@redwoodjs/auth-firebase-api`
Might be that I have to create the `firebase-admin` instance in the user's project and pass it in to the exported `authDecoder` function. I'll keep experimenting for a bit more.

> Found an easier way to make it work - make sure the auth-firebase-api package and the user's project is using the exact same version of firebase-admin. That way yarn will load the same "admin" in the decoder that's already initialized in the users's app. PR here: https://github.com/redwoodjs/redwood/pull/7122

---

There are some good comments in that PR ☝️ too that are worth reading.

With all that said I did some tests with both "your" and "my" solutions. And for me both work. And it makes sense. No matter if `admin.initializeApp()` is called in `decoder.ts` or `api/src/lib/auth.ts` it's done as a side effect of just importing the file, before any code that would use the initialized app is executed.

So I think it comes down to:
* Have the code hidden inside `decoder.ts` so that users don't have to even see it. Makes things seem less complex 
or
* Have the code in `api/src/lib/auth.ts` so that users can customize all the other options that are available for when initializing the firebase app. At the cost of more code to read and understand in their app

I still haven't decided what option I prefer. 

@rapzo Please let me know if you agree with my analysis and if you have any input of my two options

jtoar commented:
Just wanted to chime in and confirm that @Tobbe's right, I had mismatched versions of firebase-admin. Also wanted to add that I'm not sure it's worth changing where adminApp is initialized unless it's really clear that it's better in one place or the other. If we change it's not the end of the world, but it makes more breaking than simply upgrading the version of firebase-admin

rapzo commented:
Yeah, i agree too. I'm not keen in removing the availability to configure
the app from the developer (although firebase reads it from an environment
variable as well).
I'll try to look for a better solution. But I wasn't able to run it with
deduped, single version firebase-admin packages, so I'm confused.
I'll let you know as soon as I find something worth looking into.

On Mon, May 15, 2023, 22:41 Dominic Saadi ***@***.***> wrote:

> Just wanted to chime in and confirm that @Tobbe <https://github.com/Tobbe>'s
> right, I had mismatched versions of firebase-admin. Also wanted to add that
> I'm not sure it's worth changing where adminApp is initialized unless it's
> really clear that it's better in one place or the other. If we change it's
> not the end of the world, but it makes more breaking than simply upgrading
> the version of firebase-admin
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redwoodjs/redwood/pull/8305#issuecomment-1548630181>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABECTCOAG5NIZIEZ45H7JTXGKPG3ANCNFSM6AAAAAAX73D6NA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
